Parser/pegen.c

```diff
@@ -681,12 +681,50 @@ initialize_token(Parser *p, Token *token, const char *start, const char *end, in
         return -1;
     }
 
-    const char *line_start = token_type == STRING ? p->tok->multi_line_start : p->tok->line_start;
-    int lineno = token_type == STRING ? p->tok->first_lineno : p->tok->lineno;
+    const char *line_start;
+    if (token_type == STRING) {
+        line_start = p->tok->multi_line_start;
+    }
+    else if (token_type == FSTRING_MIDDLE) {
+        line_start = p->tok->tok_mode_stack[p->tok->tok_mode_stack_index].f_string_constant_multi_line_start;
+    }
+    else if (token_type == FSTRING_END) {
+        line_start = p->tok->tok_mode_stack[p->tok->tok_mode_stack_index + 1].f_string_constant_multi_line_start;
+    }
+    else {
+        line_start = p->tok->line_start;
+    }
+
+    int lineno;
+    if (token_type == STRING) {
+        lineno = p->tok->first_lineno;
+    }
+    else if (token_type == FSTRING_MIDDLE) {
+        lineno = p->tok->tok_mode_stack[p->tok->tok_mode_stack_index].f_string_constant_first_line;
+    }
+    else if (token_type == FSTRING_END) {
+        lineno = p->tok->tok_mode_stack[p->tok->tok_mode_stack_index + 1].f_string_constant_first_line;
```

Comment:
My concern is that this is accessing an invalidated stack item. For example if we decide to do something like the below [here](https://github.com/we-like-parsers/cpython/blob/435fead4bc7e6a5103bfcfe7ebb67e08ff7dcd04/Parser/tokenizer.c#L2276), this would fail.

```c
    *p_start = tok->start;
    *p_end = tok->cur-current_tok->f_string_quote_size;
     _reset_stack_item(tok, tok->tok_mode_stack_index);  // Set all the struct values to their initial values
     tok->tok_mode_stack_index--;
     return FSTRING_END;
```

Reply:
Ah, that will tell us where it ended, not where it started :(

Reply:
I am fine as long as we can somehow be confident enough that we are not going to push something meanwhile. Maybe we could add a bunch of asserts somehow.... 

@isidentical What are your thoughts?